src/ripple/nodestore/DeterministicShard.md

```diff
@@ -0,0 +1,104 @@
+# Deterministic Database Shards
+
+This doc describes the standard way to assemble the database shard. A shard assembled using this approach becomes deterministic i.e. if two independent sides assemble the shard consists of the same ledgers, accounts and transactions, then they will obtain the same shard files `nudb.dat` and `nudb.key`. The approach deals with the `NuDB` database format only, refer to `https://github.com/vinniefalco/NuDB`.
+
+
+## Headers
+
+Due to NuDB database definition, the following headers are using for database files:
+
+nudb.key:
+```
+char[8]         Type            The characters "nudb.key"
+uint16          Version         Holds the version number
+uint64          UID             Unique ID generated on creation
+uint64          Appnum          Application defined constant
+uint16          KeySize         Key size in bytes
+uint64          Salt            A random seed
+uint64          Pepper          The salt hashed
+uint16          BlockSize       Size of a file block in bytes
+uint16          LoadFactor      Target fraction in 65536ths
+uint8[56]       Reserved        Zeroes
+uint8[]         Reserved        Zero-pad to block size
+```
+
+nudb.dat:
+```
+char[8]         Type            The characters "nudb.dat"
+uint16          Version         Holds the version number
+uint64          UID             Unique ID generated on creation
+uint64          Appnum          Application defined constant
+uint16          KeySize         Key size in bytes
+uint8[64]       (reserved)      Zeroes
+```
+there all fields are saved using network byte order (most significant byte first).
+
+To make the shard deterministic the following parameters are used as values of header field both for `nudb.key` and `nudb.dat` files.
+```
+Version         2
+UID             digest(1)
+Appnum          1
+KeySize         32
+Salt            digest(0)
+Pepper          xxhasher(Salt)
+BlockSize       0x1000
+LoadFactor      0.5 (numeric 0x8000)
+```
+Note: xxhasher is native hasher of NuDB, refer to NuDB code.
+
+The `digest(i)` mentioned above defined as the follows:
+
+First, SHA256 hash `H` calculated of the following structure
+```
+uint256         lastHash        Hash of last ledger in shard
+uint32          index           Index of the shard
+uint32          firstSeq        Sequence number of first ledger in the shard
+uint32          lastSeq         Sequence number of last ledger in the shard
+uint32          version         Version of shard, 2 at the present
+```
+
+Then, `digest(i)` is defined as `i`-th 64-bit portion of the above hash `H` with swapped order of bytes, i.e.
+```
+digest(i) = H[i*8+0] << 56 | H[i*8+1] << 48 | ... | H[i*8+7] << 0.
+```
+where `H[i]` denotes `i`-th byte of hash `H`.
+
+
```

Comment:
First question, and this is more theoretical: should we modify `NuDB` to increase the size of the headers of the key and data files so that they both are _precisely_ 512 bytes, increase the size of some fields (specifically the UID field that should be 256 bits) and allocate a portion of the header as an area of the database owner to store his own metadata? 

Now on to more practical matters: I really don't like the way that we're taking a cryptographic hash function, chopping its output up and manipulating it.

I recommend the following: replace `SHA256` with `RIPEMD-160` and use it to calculate `H` over the following data:

```
uint256         lastHash        Hash of last ledger in shard
uint32          index           Index of the shard
uint32          firstSeq        Sequence number of first ledger in the shard
uint32          lastSeq         Sequence number of last ledger in the shard
uint32          version         Version of shard, 2 at the present
```

Crack the 160 bit hash `H` apart into **2** 64-bit values `A` and `B` and **1** 32-bit value `C` and set: 

1. `UID := A`;
2. `SALT := B`; and
3. `APPNUM := C | 0x5348524400000000`. (`0x53585244 == 'S' 'H' 'R' 'D'`).

Reply:
Saving hash of shard as UID/salt was not required by this ticket. It is my proposition. Ticket recommends to set UID/salt to known constant + shard_index. Your proposition of using RIPEMD-160 instead of SHA256 is better. I think that changing NuDB standard is much more difficult because current standard is used is main net if I am not mistaking.

Reply:
Done